web/js/codeworld-mode.js

```diff
@@ -82,43 +116,91 @@ CodeMirror.defineMode('codeworld', (_config, modeConfig) => {
 
         if (stream.match(RE_DASHES)) {
             stream.skipToEnd();
-            return 'comment';
+            return ['comment'];
         }
 
-        if (stream.match(RE_QUAL)) return 'qualifier';
-        if (stream.match(RE_VARID) || stream.match(RE_VARSYM)) return 'variable';
-        if (stream.match(RE_CONID) || stream.match(RE_CONSYM)) return 'variable-2';
-        if (stream.match(RE_NUMBER)) return 'number';
-        if (stream.match(RE_CHAR) || stream.match(RE_STRING)) return 'string';
+        if (checkContextStarter(stream, state)) {
+            return ['keyword'];
+        }
+
+        if (state.scanState === NORMAL && stream.match(RE_IN)) {
+            // Cancel previous layout 'let' context
+            state.layoutContext.pop();
+            return ['keyword'];
+        } else if (state.scanState === NORMAL) {
+            let lastContext = state.layoutContext[state.layoutContext.length - 1];
+            lastContext = lastContext === '{' ? 0 : lastContext;
+            if (stream.sol() && stream.indentation() < lastContext) {
+                // Close previous context
+                state.layoutContext.pop();
+            }
+            if (stream.column() === lastContext || stream.column() === 0) {
+                spanStyles.push('layout');
+            }
+        } else if (state.scanState === TRIGGERED && stream.match('{')) {
+            // Cancel previous layout context because of next char is '{'
+            state.layoutContext.pop();
+            state.scanState = NORMAL;
+        } else if (state.scanState === TRIGGERED) {
+            state.layoutContext.push(stream.column());
+            spanStyles.push('layout');
+            state.scanState = NORMAL;
+        }
+
+        if (stream.match(RE_QUAL)) {
+            spanStyles.push('qualifier');
+            return spanStyles;
+        }
+        if (stream.match(RE_VARID) || stream.match(RE_VARSYM)) {
+            spanStyles.push('variable');
+            return spanStyles;
+        }
+        if (stream.match(RE_CONID) || stream.match(RE_CONSYM)) {
+            spanStyles.push('variable-2');
+            return spanStyles;
+        }
+        if (stream.match(RE_NUMBER)) {
+            spanStyles.push('number');
+            return spanStyles;
+        }
+        if (stream.match(RE_CHAR) || stream.match(RE_STRING)) {
+            spanStyles.push('string');
+            return spanStyles;
+        }
 
         if (stream.match(RE_OPENBRACKET)) {
-            state.brackets.push(stream.current());
-            return `bracket${state.brackets.length <= 7 ? `-${  
-                state.brackets.length - 1}` : ''}`;
+            state.layoutContext.push(stream.current());
+            spanStyles.push(
+                `bracket${state.layoutContext.length <= 7 ? `-${
```

Comment:
I think the right thing to do here is count the explicit brackets in the layout context.  You should not bump the bracket count when there's an implicit layout.